pkg/controllers/user/pipeline/controller/pipelineexecution/logsyncer.go

```diff
@@ -0,0 +1,76 @@
+package pipelineexecution
+
+import (
+	"context"
+	"fmt"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/engine"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/utils"
+	"github.com/rancher/rancher/pkg/ticker"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+const (
+	syncLogInterval = 20 * time.Second
+)
+
+type ExecutionLogSyncer struct {
+	pipelineExecutionLister    v3.PipelineExecutionLister
+	pipelineExecutionLogLister v3.PipelineExecutionLogLister
+	pipelineExecutionLogs      v3.PipelineExecutionLogInterface
+	cluster                    *config.UserContext
```

Comment:
please ask alena if we can keep cluster in the controller.

Reply:
As I've mentioned above, it is better to clearly define listers and interfaces rather than putting the generic cluster object in.